modules/python: Add a find_module method

dcbaker opened the pull request:
This adds a new method for querying the availability of third party
python modules available. This is designed to provide information in
the configuraiton log while providing the familiar dependency
interface within meson itself.

Fixes #2377

ignatenkobrain commented on mesonbuild/modules/python.py:
```diff
@@ -461,6 +463,71 @@ def get_variable_method(self, args, kwargs):
+
+        module = PythonPackage(name, self.interpreter.environment, None,
+                               {'found': found, 'version': found_version})
+        
```
__[Flake8]__ [W293] blank line contains whitespace ([link](https://sider.review/gh/repos/19784232/pulls/5204#issue-3944732))

<sub>posted by Sider</sub>

lazka commented:
Any reason to not check on a package level instead which already has a version attached and doesn't require importing?

dcbaker commented:
pkg_resources is part of setuptools and we have a strict no dependencies policy. I guess whenever we get to a hard dependency on 3.7 we could use `importlib.resources`, but that's a ways off. I guess I should also pick a module that's more reliably installed than numpy.

lazka commented:
Note that this is a dependency of the Python you are building for, not meson itself.

dcbaker commented:
I'm not sure I understand what your saying?

lazka commented:
> I'm not sure I understand what your saying?

If you need to find dependencies that are using pkg-config you need to call pkg-config, if you need to find deps using cmake you need to call cmake, if you need to find python deps you need to call pkg_resources...

Anyway, I don't feel that strongly about it, just wanted to point out that native dependency handling in Python usually happens at a different layer. The module level one is of course still useful.

dcbaker commented:
Okay, that is a valid point. We could probably handle this the same way we do dependencies, with different finders. Which is probably better.

dcbaker commented:
the test failure is spurious

nirbheek commented:
> pkg_resources is part of setuptools and we have a strict no dependencies policy

Funny story, we depend on `pkg_resources` now because it's available on all platforms with Python *except* Linux distros that have split it out into its own package.

dcbaker commented:
@jpakkane, opinions?

jpakkane commented:
First of all it should not be called `has_module`, because a function of that name should return a boolean, not a dependency object.

As for the dep, it would be nice if we could do this without extra dependencies. That being said since this is strictly optional and only used via process invocation on a different Python I guess we could accept this, if it makes things a lot simpler and/or faster. There needs to be a comment describing the issue and unambiguously saying that this is only acceptable here, not in Meson proper.

dcbaker commented:
I renamed it `find_module` already, I feel like it should be analogous to `compiler.find_library`, we could even call it `find_library` if you think that would make more sense.

If we added support for `pkg_resources` I'd like to do it in the same way we handle `pkg-config` for C, as one option that can fall back to a the "read this attribute" method.

jpakkane commented on docs/markdown/Python-module.md:
````diff
@@ -219,6 +219,42 @@ a non-existing variable will cause a fatal error.
+```
+
+This method works much like the `dependency` function, including
+returning an object that wokrs like a dependency, including the
````
"wokrs"

eli-schwartz commented:
> pkg_resources is part of setuptools and we have a strict no dependencies policy.

Isn't pkg_resources a bit special inasmuch as the actual meson command is a setuptools-generated stub that requires pkg_resources ?

jpakkane commented:
It is unclear what the dependency object _actually_ contains. If you do a lookup for Numpy, does the dependency object have things needed to link against Numpy libraries or is it the same as the Python dependency?

dcbaker commented:
Its the same as the python dependency. The reason I made it dependency like is so that it can be used as a disabler.

jpakkane commented:
The commit has a few `has_module`s that should be `find_module`s.

I see how this naming would work but it still feels a bit off. How about an alternative approach of adding a kwarg to `dependency` instead? So something in the vein of:

```meson
py_dep = py_installation.dependency(..., require_modules: ['setuptools'])
```

dcbaker commented:
What if the modules are optional, or, like mesa you need a module like mako to use a in custom_target/generator but you don't want to link with python, since `py_installation.dependency()` is for linking with python, correct?

lazka commented:
(drive by comment: a successor to pkg_resource was merged into CPython (3.8) just now: https://github.com/python/cpython/pull/12547/files)

jpakkane commented:
> you need a module like mako to use a in custom_target/generator but you don't want to link with python

How would you use this functionality do achieve that, specifically? The return value of both is the same, a Python dependency. I don't see how they would be used differently.

dcbaker commented:
I can see advantages to going both ways, the `require_modules` approach would allow us to theoretically check through many versions of python if you said 'python3', and had 3.6 and 3.7 and only 3.6 has the modules you need. The other one allows you to have optional modules and check for those. For mesa `require_modules` would be sufficient.

jpakkane commented:
So can we go with adding `require_modules` only then? I would imagine that to be the common case. If people need something more advanced we can later add a second method for that (and then we will have real world usage example of what we specifically need to do).

jpakkane commented:
Since this is almost done and useful, it gets a dev freeze exception.

dcbaker commented:
Sure, I'll make the changes.

ignatenkobrain commented on mesonbuild/modules/python.py:
```diff
@@ -528,13 +533,38 @@ def find_installation(self, interpreter, state, args, kwargs):
+        if python.found() and want_modules:
+            for mod in want_modules:
+                p, out, err = mesonlib.Popen_safe(
+                    python.command + \
```
__[Flake8]__ [E502] the backslash is redundant between brackets ([link](https://sider.review/gh/repos/19784232/pulls/5204#issue-4191315))

<sub>posted by Sider</sub>

dcbaker commented:
Okay, I've completely reworked this with @jpakkane's suggestion to extend the signature of `find_installation` instead of adding a new method. This does limit us a bit in that we can't do version checking, but it is a much simpler implementation than the original proposal. I have chosen to use `modules` instead of `require_modules`, as we already have precedent for using that name in the `dependency` function.

Ericson2314 commented on mesonbuild/modules/python.py:
```diff
@@ -491,11 +492,15 @@ def _check_version(self, name_or_path, version):
     @FeatureNewKwargs('python.find_installation', '0.49.0', ['disabler'])
+    @FeatureNewKwargs('python.find_installation', '0.51.0', ['modules'])
     @disablerIfNotFound
-    @permittedKwargs(['required'])
```
orthogonal, but if we allow `native` here and pass it on to `ExternalProgram` I think it will work?

dcbaker replied:
I guess? I don't know why you'd want to find a python executable for !build though?

Ericson2314 replied:
@dcbaker then `ExternalProgram` should be given `MachineChoice.build` unconditionally as it defaults to the host platform like everything else I think.
